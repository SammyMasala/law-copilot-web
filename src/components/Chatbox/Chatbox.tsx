import React, { useContext, useEffect, useRef, useState } from "react";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import ListGroup from "react-bootstrap/ListGroup"
import Form from "react-bootstrap/Form"
import Button from "react-bootstrap/Button"
import { ChatService } from "@src/services";
import { ChatMessage } from "@src/entities";

interface IChatboxProps{
    context: React.Context<any>
}
    
export const Chatbox: React.FC<IChatboxProps> = (props: IChatboxProps) => {
    const chatService = new ChatService();
    const {context} = props
    const { messages, setMessages } = useContext(context)
    const [isInputEnabled, setIsInputEnabled] = useState<boolean>(true)
    const [input, setInput] = useState<string>("");
    const inputRef = useRef<HTMLInputElement>(null)

    useEffect(() => {
        const chatboxContainer = document.querySelector("#chatbox-messages") as HTMLElement
        if (chatboxContainer) {
            chatboxContainer.style.height = `${chatboxContainer.offsetHeight.toString()}px`;
        }
    }, [])

    // Disable chatbox while waiting for reply
    useEffect(() => {
        if (!messages.length || !messages.at(-1).isUser) {
            return
        }
        const sendMessage = async (messages: ChatMessage[]) => {
            const reply = {
                message: await chatService.chatQuery(messages),
                isUser: false
            }
            setMessages([...messages, reply])
        }

        if (messages.length && messages.at(-1)?.isUser) {
            setIsInputEnabled(prev => false)
            sendMessage(messages).catch(err => {
                console.error(err)
            }).then(() => {
                setIsInputEnabled(prev => true)
            })
        }
    }, [messages])

    useEffect(() => {
        if (isInputEnabled) {
            inputRef!.current!.disabled = false;
            inputRef!.current!.placeholder = "Ask about..."
        } else {
            inputRef!.current!.disabled = true;
            inputRef!.current!.placeholder = "...awaiting reply..."
        }
    }, [isInputEnabled])

    const handleChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;
        setInput(value)
    }

    const handleSubmitInput = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault()
        const message = {
            message: input,
            isUser: true
        }
        setMessages([...messages, message])
        setInput("")
    }
    return (
        <Container className="d-flex flex-column">
            <Row id="chatbox-messages" className="flex-grow-1 overflow-auto">
                <ListGroup>
                    {messages.map((message: ChatMessage, index: number) => {
                        return (
                            <ListGroup.Item
                                className={`d-flex border border-0 bg-transparent
                                ${message.isUser === true ? "justify-content-end" : "justify-content-start"}`}
                                key={`list-${index}`}
                            >
                                <Button
                                    variant={`${message.isUser === true ? "dark" : "secondary"}`}
                                    style={
                                        {
                                            wordBreak: "break-all",
                                            overflowWrap: "break-word",
                                            textAlign: "justify"
                                        }
                                    }
                                >
                                    {message.message}
                                </Button>
                            </ListGroup.Item>
                        )
                    })}
                </ListGroup>

            </Row>
            <Row id="chatbox-input">
                <Form className="d-flex" onSubmit={handleSubmitInput}>
                    <Form.Control
                        type="text"
                        name="input"
                        ref={inputRef}
                        value={input}
                        aria-describedby="user message input block"
                        onChange={handleChangeInput}
                    />
                    <Button variant="dark" type="submit">{"Send"}</Button>
                </Form>
            </Row>
        </Container>
    )
}