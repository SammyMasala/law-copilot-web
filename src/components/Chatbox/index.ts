export * from "./Chatbox"
export * from "./types"