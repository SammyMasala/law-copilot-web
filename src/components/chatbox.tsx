import React, { FormEvent, useState } from 'react';
import Form from 'react-bootstrap/Form';
import ListGroup from 'react-bootstrap/ListGroup'
import Button from 'react-bootstrap/Button';
import Image from 'react-bootstrap/Image';
import { getMessageResponse } from '../api';

import bot from '../static/bot.png';
import user from '../static/user.png';

interface IMessage{
    text: string,
    isUser: boolean
}

interface IChatBoxInput{
    messageInput: string
}


const ChatBox: React.FC = () => {
    const [messages, setMessages] = useState<IMessage[]>([])
    const [formData, setFormData] = useState<IChatBoxInput>({messageInput: ""})

    const handleFormSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        try{
            const newMessage: IMessage = {
                text: formData.messageInput,
                isUser: true
            }
            setMessages(prevState => ([...prevState, newMessage]))

            // Reset form
            setFormData({messageInput: ""})

            // Send message, await message reply
            const reply: IMessage = {
                text: await getMessageResponse(formData.messageInput),
                isUser: false
            }

            setMessages(prevState => ([...prevState, reply]))   
            console.log(messages)         
        }catch(err){
            console.error(err)
            handleFormSubmitError()
        }
        
    }

    const handleFormChangeUserInput = (e: React.ChangeEvent<HTMLInputElement>) => {
        const {name, value} = e.target
        setFormData({
            ...formData,
            [name]: value
        });
    };

    const handleFormSubmitError = () => {
        //TODO: Show message notifying user of service error near input area
    }
    
    return (
        <>
        <ListGroup>
            {messages.map((message, index) => {
                return (
                    <ListGroup.Item id={"message" + index.toString()}>
                        <h4>{message.text}</h4>
                        <Image src={message.isUser === true ? user : bot} />
                    </ListGroup.Item>
                )
            })}
        </ListGroup>
        <Form onSubmit={handleFormSubmit}>
            <Form.Group controlId='formTextInput'>
                <Form.Control 
                type='text' 
                placeholder='Insert message here...' 
                aria-describedby='userInputField' 
                name='messageInput' 
                value={formData.messageInput}
                onChange={handleFormChangeUserInput}              
                />
                <Button type='submit'>Submit</Button>
            </Form.Group>
        </Form>
        </>
    )
}

export default ChatBox;