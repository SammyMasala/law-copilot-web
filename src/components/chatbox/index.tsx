import React, { useEffect, useState } from "react";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import ListGroup from "react-bootstrap/ListGroup"
import { getMessageResponse } from "../../api";

import MessageInput from "./messageInput";
import ChatBubble from "./chatBubble";

import "./styles.css";
import type { 
    IChatBoxProps, 
    IChatBubbleStyle, 
    IMessage, 
    ISelectedButton 
} from "./interfaces";
import type { 
    IChatBoxProps, 
    IChatBubbleStyle, 
    IMessage, 
    ISelectedButton 
} from "./interfaces";

const ChatBox: React.FC<IChatBoxProps> = ({setSelectedButtonText, messagesToAppend, setCurrentMessages}) => {
    const [messages, setMessages] = useState<IMessage[]>([])

    // Append Past Messages
    useEffect(() => {
        const appendMessagesToChatbox = () => {
            setMessages([...messagesToAppend, ...messages])                
        }

        if(messagesToAppend.length){
            appendMessagesToChatbox()
        }
    }, [messagesToAppend])

    // Pass Messages to parent
    useEffect(() => {
        if(messages.length){
            setCurrentMessages(messages)
        }
    }, [messages])

    const handleFormSubmit = async (messageInput: string) => {
        try{
            const newMessage: IMessage = {
                text: messageInput,
                isUser: true
            }
            setMessages(prevState => ([...prevState, newMessage]))

            // Send message, await message reply
            const reply: IMessage = {
                text: await getMessageResponse(messageInput),
                isUser: false
            }

            setMessages(prevState => ([...prevState, reply]))   
        }catch(err){
            console.error(err)
            handleFormSubmitError()
        }
        
    }

    const handleSelectButton = (selected: ISelectedButton) => {
        setSelectedButtonText(selected.message)
    }

    const handleFormSubmitError = () => {
        //TODO: Show message notifying user of service error near input area
    }
    
    return (
        <Container fluid className="bg-dark bg-opacity-25 d-flex h-100 p-1 flex-column">
            <Row className="flex-grow-1 m-1 p-1 overflow-auto">
                <ListGroup>
                    {messages.map((message, index) => {
                        const bubbleStyle: IChatBubbleStyle = {
                            variant: message.isUser === true ? "dark": "secondary",
                            className: "button-wrap text-start"
                        }
                        return (  
                            <ListGroup.Item 
                                id={`list-item-${index.toString()}`}
                                className={`border border-0 bg-transparent d-flex ${message.isUser === true ? "justify-content-end": "justify-content-start"}`}
                            >                                    
                                <ChatBubble 
                                    selectTextBubble={handleSelectButton} 
                                    message={message} 
                                    bubbleStyle={bubbleStyle}
                                />
                            </ListGroup.Item>                        
                        )
                    })}
                </ListGroup>
            </Row>            
            <Row className="mb-1 p-1">
            <Row className="mb-1 p-1">
                <MessageInput 
                    submitMessage={handleFormSubmit}
                />
            </Row>            
        </Container>
    )
}

export default ChatBox;
export { IMessage, ISelectedButton }