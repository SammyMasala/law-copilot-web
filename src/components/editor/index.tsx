import React, { useEffect, useState } from "react"
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Image from "react-bootstrap/Image"
import Button from "react-bootstrap/Button";
import ReactQuill, { EmitterSource } from "react-quill-new"; 
import Delta from "quill-delta"; 

import "react-quill-new/dist/quill.snow.css";

interface IEditorProps{
    onChange: (document: string) => void
}
const Editor: React.FC<IEditorProps> = ({onChange}) => {
    const [documentHTML, setDocumentHTML] = useState<string>("")

    // Chain update to parent
    useEffect(() => {
        onChange(documentHTML)
    },[documentHTML])

    const handleChangeDocument = (value: string, delta: Delta, source: EmitterSource, editor: ReactQuill.UnprivilegedEditor) => {
        setDocumentHTML(editor.getHTML())
    } 


    return (
        <Container>
            <Row id="editor-save"></Row>
            <Row id="editor-quill">
                <ReactQuill theme="snow" onChange={handleChangeDocument}/>
            </Row>
        </Container>
    )
} 

export default Editor; 