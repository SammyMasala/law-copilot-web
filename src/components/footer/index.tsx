import React from "react";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row"

interface IFooter {
    className: string
}

const Footer: React.FC<IFooter> = ({className}) => {
    return (
        <Container className={className}>
            <Row>
                <a href="https://github.com/SammyMasala/law-copilot-webs">GitHub Page</a>
            </Row>
        </Container>
    )
} 

export default Footer;