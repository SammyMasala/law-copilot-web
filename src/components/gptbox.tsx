import React from "react";
import { Card, CardBody } from "react-bootstrap";

const GPTBox: React.FC = () => {
    return (
        <Card style={{ width: '18rem'}}>
            <Card.Body>
                <Card.Title>GPT Section-WIP</Card.Title>
                <Card.Text>This is where the chat should be</Card.Text>
            </Card.Body>

        </Card>
    )
}

export default GPTBox;