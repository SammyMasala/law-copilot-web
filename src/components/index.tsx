export * from "./gptbox";
export * from "./quill";