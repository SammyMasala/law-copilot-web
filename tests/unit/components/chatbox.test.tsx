import React, { createContext, useState } from "react";
import "@testing-library/jest-dom"
import {render, screen, cleanup} from "@testing-library/react"
import Chatbox from "../../../src/components/Chatbox";
import { IMessage } from "../../../src/components/Chatbox";

afterEach(()=> {
    cleanup();
})

describe("Header Component", () => {
    const mockContext:React.Context<any> = createContext(null)
    const mockMessages:IMessage[] = [{message: "test message", isUser: false}]
    const MockContextProvider = ({children}: any) => {
        const [messages, setMessages] = useState<IMessage[]>(mockMessages)
        const [isLoaded, setIsLoaded] = useState<boolean>(true)
        return (
            <mockContext.Provider 
                value={{messages, setMessages, isLoaded}}>
                    {children}
            </mockContext.Provider>
        )
    }


    test("initial state", async () => {
        const TestComponent:React.FC = () => {
            return (
                <MockContextProvider>
                    <Chatbox context={mockContext}/>
                </MockContextProvider>
            )
        }     
    
        render(<TestComponent />)
        const title = screen.getByText("test message")
        expect(title).toBeInTheDocument();
    })

    test("snapshot", () => {   
        const TestComponent:React.FC = () => {
            return (
                <MockContextProvider>
                    <Chatbox context={mockContext}/>
                </MockContextProvider>
            )
        }     
    
        const {asFragment} = render(<TestComponent />)
        expect(asFragment()).toMatchSnapshot();
    })
})