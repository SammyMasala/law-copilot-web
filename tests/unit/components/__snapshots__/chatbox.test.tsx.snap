// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column container"
  >
    <div
      class="flex-grow-1 overflow-auto row"
      id="chatbox-messages"
      style="height: 0px;"
    >
      <div
        class="list-group"
      >
        <div
          class="d-flex border border-0 bg-transparent
                                justify-content-start list-group-item"
        >
          <button
            class="btn btn-secondary"
            style="word-break: break-all; overflow-wrap: break-word; text-align: justify;"
            type="button"
          >
            test message
          </button>
        </div>
      </div>
    </div>
    <div
      class="row"
      id="chatbox-input"
    >
      <form
        class="d-flex"
      >
        <input
          aria-describedby="user message input block"
          class="form-control"
          name="input"
          placeholder="Ask about..."
          type="text"
          value=""
        />
        <button
          class="btn btn-dark"
          type="submit"
        >
          Send
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
